### ISSUE TITLE
Add xADComputer

### ISSUE THREAD

**Issue by gmatusko** [PowerShell/xActiveDirectory#69]
I'm using DSC to create clusters and it would be nice/easier if there was a xADComputer to add computer objects to the domain ahead of creating the cluster.  Suggested parameters would be OU location and account Enabled = True/False.  Looking for the Enabled parameter as cluster computer objects need to be disabled before creating the cluster.

**Comment by iainbrighton** [comment]
Sounds like a legitimate request!

What about a generic xADObject resource instead? Might be handy for moving any object too.

It could be a precursor to a resource that can set permissions on an AD object... ;)

**Comment by gmatusko** [comment]
I like the sounds of that but how would you account for all the various properties for AD objects the command would cover?  If you are thinking about any AD object that is, not just computer objects.

**Comment by PlagueHO** [comment]
I can see the benefit in this too.

As well as an xADManagedServiceAccount/xADGroupManagedServiceAccount too (while we're on the topic).

**Comment by iainbrighton** [comment]
@gmatusko By using the *-ADObject cmdlets to create a computer object (or any other object type) :wink:

**Comment by PlagueHO** [comment]
@iainbrighton - although I guess what @gmatusko means is that because the different types of objects have different property requirements, which properties would be exposed in the resource schema.

Specifically, a way of supporting the "OtherAttributes" parameter on the *-ADObject cmdlet would be awesome - but I've not had a lot of luck implementing any kind of "Hashtable" like property on a resource (although I may have just been doing things wrong).

Also, hijacking the idea, implementing MSA/gMSA using *-ADObject is possible but I think probably dangerous as it would depend on the user knowing which properties of the ADObject to set for a MSA/gMSA. Same would go for other less well used types.

Just thinking out loud here really :smile:

**Comment by iainbrighton** [comment]
@gmatusko I wasn't planning on implementing all properties; just a DN, object class and an enabled property would have made for a simple generic resource. That was until @PlagueHO waded in with his size 10's :stuck_out_tongue_closed_eyes:

Could we cover computer accounts, managed service accounts and group managed service accounts with a single resource?

**Comment by PlagueHO** [comment]
LOL :smile: More like size 11's! 

I think maybe we should just leave gMSA/MSA out of this. It would be a nice to have one day but no point in slowing more important stuff down with it.

You know what would be really nice? Being able to subclass DSC resources. Then we could create gMSA/MSA resources by subclassing the computer resource (they are after all a subclass of the computer AD class). I wonder if this is possible given that you can define DSC resources using a class structure... sorry, a bit far off topic I'm sure.

**Comment by PlagueHO** [comment]
Or perhaps an xADObject could be implemented and also an xADObjectProperty as well which could be used to set the more esoteric properties required in some situations (e.g. MSA user bindings and what-not)?

**Comment by iainbrighton** [comment]
@gmatusko @PlagueHO Having looked into this in more detail, I'm just gonna stick with the xADComputer resource! Creating a Computer object, for example, does lots behind the scenes including placing the object in the 'Domain Computers' group and setting various attributes. I'm not saying having a generic isn't worthwhile - it just needs thinking about a bit more.

**Comment by PlagueHO** [comment]
@iainbrighton - completely - seems like a generic might create more problems than it solves.

**Comment by GoateePFE** [comment]
Agreed. Sounds like object-specific resources are the way to go for now. Eventually I would like to see a generic AD object resource as a catch-all for other things.

**Comment by iainbrighton** [comment]
@PlagueHO @GoateePFE Here is what I have so far. Is there anything else you think needs to be implemented?:
```powershell
[ClassVersion("1.0.0.0"), FriendlyName("xADComputer")] 
class MSFT_xADComputer : OMI_BaseResource
{
    [Key, Description("Specifies the name of the computer")] String ComputerName;
    [Write, Description("Specifies the location of the computer, such as an office number")] String Location;
    [Write, Description("Specifies the fully qualified domain name (FQDN) of the computer")] String DnsHostName;
    [Write, Description("Specifies the service principal names for the computer account")] String ServicePrincipalNames[];
    [Write, Description("Specifies the UPN assigned to the computer account")] String UserPrincipalName;
    [Write, Description("Specifies the display name of the computer")] String DisplayName;
    [Write, Description("Specifies the X.500 path of the Organizational Unit (OU) or container where the computer is located")] String Path;
    [Write, Description("Specifies a description of the computer object")] String Description;
    [Write, Description("Specifies if the computer account is enabled")] Boolean Enabled;
    [Write, Description("Specifies the user or group Distinguished Name that manages the computer object")] String Manager;
    [Write, Description("Specifies the Active Directory Domain Services instance to connect to perform the task")] String DomainController;
    [Write, Description("Specifies the user account credentials to use to perform the task"), EmbeddedInstance("MSFT_Credential")] String DomainAdministratorCredential;
    [Write, ValueMap{"Present", "Absent"},Values{"Present", "Absent"}] String Ensure;
    [Read, Description("Returns the X.500 path of the computer object")] String DistinguishedName;
    [Read, Description("Returns the security identifier of the computer object")] String SID;
};

```
```powershell
xADComputer [String] #ResourceName
{
    ComputerName = [string]
    [Description = [string]]
    [DisplayName = [string]]
    [DnsHostName = [string]]
    [DomainAdministratorCredential = [PSCredential]]
    [DomainController = [string]]
    [Enabled = [bool]]
    [Ensure = [string]{ Absent | Present }]
    [Location = [string]]
    [Manager = [string]]
    [Path = [string]]
    [ServicePrincipalNames = [string[]]]
    [UserPrincipalName = [string]]
}
```

**Comment by gmatusko** [comment]
Iain,

Configuring delegation would be helpful as well but not sure how difficult that would be to implement.  I see you are covering SPN which is great but the ability to add an SPN to the delegation tab would be very helpful in our environments.  I wouldn't hold up the release for this but something to think about for the future.

Greg


________________________________

**Comment by PlagueHO** [comment]
@iainbrighton: Looks pretty good to me. Nice one!
@gmatusko: Implementing the delegation would be nice but from a purely technical standpoint I reckon we'd probably need to implement delegation in it's own resource - so we could do this:
```powershell
xADComputerDelegate Add_DelegateCifsApp1a
{
    ComputerName = 'SA-APP1'
    Ensure = 'Present'
    SPN = 'cifs\SA-FS1'
}
xADComputerDelegate Add_DelegateCifsApp1b
{
    ComputerName = 'SA-APP1'
    Ensure = 'Present'
    SPN = 'cifs\SA-FS1.CONTOSO.COM'
}
xADComputerDelegate Remove_DelegateCifsApp2a
{
    ComputerName = 'SA-APP2'
    Ensure = 'Absent'
    SPN = 'cifs\SA-FS1'
}
xADComputerDelegate Remove_DelegateCifsApp2b
{
    ComputerName = 'SA-APP2'
    Ensure = 'Absent'
    SPN = 'cifs\SA-FS1.CONTOSO.COM'
}
```
My 2c :smile:

**Comment by PlagueHO** [comment]
Soooo @iainbrighton, how is xADComputer coming along :smile: Now that TP5 is out and unfortunately there is still no was of domain joining a Nano Server, we'll need to implement ODJ blob support in xADComputer. So once you've got xADComputer in place I'll then go about adding the ODJ support to it. No hassle though - I've got a more than enough to keep my busy at the moment :+1:

**Comment by iainbrighton** [comment]
Weeeeeeeeeeell @PlagueHO, I've nearly finished the test coverage! I still need to update the readme.

I've pushed my latest commits [upstream](https://github.com/iainbrighton/xActiveDirectory/tree/Issue69/DSCResources/MSFT_xADComputer) if you want to review before I submit a PR?

**Comment by PlagueHO** [comment]
@iainbrighton - cool! Thank you sir! I'm checking through your branch now and will let you know if I see anything untoward. It will also give me an opportunity to plan what would need to be added to support the ODJ. Cheers!

**Comment by PlagueHO** [comment]
Just a couple of things (and I am aware that they are extremely nitpicky :smile: ) :
1. Property spelled wrong in comment (https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L92).
2. These lines are > 100 chars - could we split them?
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L222
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L237
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L251
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L257
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L370
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L394
3. Comment not quite correct as Common Name isn't being set here:
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L228
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L330

Other than those tweaks above, everything looks very nice indeed!

I also think I've got a fairly good handle on how this might be changed to support ODJ. Only one parameter would need to be added, and that would be the path to the ODJ file to create. Then only this line would need to be changed:
https://github.com/iainbrighton/xActiveDirectory/blob/Issue69/DSCResources/MSFT_xADComputer/MSFT_xADComputer.psm1#L345
If the ODJ path was set then a call to DJOIN could be made, otherwise the existing process would remain. The code following L345 would remain exactly as it is.

**Comment by iainbrighton** [comment]
@PlagueHO No problem with being nitpicky - at least someone picked up on my copy'n'paste errors! Thanks for the feedback - I think I've addressed it all :smile: 

Let me know if there's anything else needed. I'm not sure I've covered all the test cases yet, e.g. ServicePrincipalNames.. 

Once this is done it'll be on all the required updates to xADDomainController...

PS - Who's idea is 100 chars?! With indentation and no positional parameters, it's almost impossible to maintain (and I personally dislike to use the backtick) :-1:

**Comment by PlagueHO** [comment]
@iainbrighton - I know what you mean about the 100 chars. Looking at the blame it was @KarolKaczmarek :smile: I understand the reasoning though - long lines can be a PITA for reviewers and just looking at code and I guess there needed to be a cut off somewhere.

I'll take a better look at your unit test coverage and see if I can suggest any tests to add (especially around the SPN). Thanks again for this!

**Comment by PlagueHO** [comment]
@iainbrighton - sorry I still haven't got around to reviewing this yet! :frowning:
